### PR TITLE
New organisation of algorithms

### PR DESCRIPTION
**Pull request by rrsadykov**

### CODE REVIEW

**Comment by rrsadykov** [comment]
All tests are ok. Ready to be merged for me.

File: src/Algorithm/branching/branchingalgo.jl [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        push!(slaves, (reform, typeof(phase.conquer_algo)))
+        getslavealgorithms!(phase.conquer_algo, reform, slaves)
+    end
+    for prioritised_rule in algo.rules
+        push!(slaves, (reform, typeof(prioritised_rule.rule)))
+        getslavealgorithms!(prioritised_rule.rule, reform, slaves)
+    end
+end
+
+# stopped here
```
Comment (guimarqu) [inline src/Algorithm/branching/branchingalgo.jl]: ?

File: test/unit/algorithms/masteripheur.jl [hunk 2]
```diff
@@ -4,22 +4,22 @@ end
 
 #CL.to_be_pruned(n::CL.Node) = true # issue 166
 
-struct InfeasibleMasterIpHeur <: ClA.AbstractConquerStrategy end
+# struct InfeasibleMasterIpHeur <: ClA.AbstractConquerAlgorithm end
```
Comment (guimarqu) [inline test/unit/algorithms/masteripheur.jl]: rm

File: test/unit/algorithms/algorithm.jl [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
 
 function algorithm_unit_tests()
     #algorithm_fallbacks_tests()
     masteripheur_tests()
 end
 
 function algorithm_fallbacks_tests()
-    @test_throws ErrorException CL.prepare!(ClA.UnknownAlgo(), nothing, nothing)
-    @test_throws ErrorException CL.run!(ClA.UnknownAlgo(), nothing, nothing)
+    # @test_throws ErrorException CL.prepare!(ClA.UnknownAlgo(), nothing, nothing)
```
Comment (guimarqu) [inline test/unit/algorithms/algorithm.jl]: rm

File: test/runtests.jl [hunk 4]
```diff
@@ -26,6 +26,8 @@ include("pricing_callback_tests.jl")
 
 rng = MersenneTwister(1234123)
 
+#mytest()
```
Comment (guimarqu) [inline test/runtests.jl]: rm

File: src/Algorithm/treesearch.jl [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+"""
+    TreeSearchAlgorithm
+
+    This algorithm uses search tree to do optimization. At each node in the tree, we apply
+    conquer algorithm to improve the bounds and divide algorithm to generate child nodes.
+"""
+Base.@kwdef struct TreeSearchAlgorithm <: AbstractOptimizationAlgorithm
+    conqueralg::AbstractConquerAlgorithm = ColGenConquer()
+    dividealg::AbstractDivideAlgorithm = SimpleBranching()
```
Comment (guimarqu) [inline src/Algorithm/treesearch.jl]: maybe only `conquer` & `divide`

File: src/Algorithm/strategies/strategy.jl [hunk 6]
```diff
@@ -1,76 +1,76 @@
-"""
-    AbstractConquerStrategy
+# """
```
Comment (guimarqu) [inline src/Algorithm/strategies/strategy.jl]: rm file

File: src/Algorithm/strategies/divide/simplebranching.jl [hunk 7]
```diff
@@ -1,4 +1,4 @@
-struct NoBranching <: AbstractDivideStrategy end
-function apply!(strategy::NoBranching, reform::Reformulation, node::Node)
-    return
-end
+# struct NoBranching <: AbstractDivideStrategy end
```
Comment (guimarqu) [inline src/Algorithm/strategies/divide/simplebranching.jl]: rm file

File: src/Algorithm/strategies/conquer/simplebnp.jl [hunk 8]
```diff
@@ -1,81 +1,81 @@
-struct SimpleBnP <: AbstractConquerStrategy 
-    colgen::ColumnGeneration
-    mastipheur::MasterIpHeuristic
-end
+# struct BasicColGenConquer <: AbstractConquerAlgorithm
```
Comment (guimarqu) [inline src/Algorithm/strategies/conquer/simplebnp.jl]: rm file

File: src/Algorithm/strategies/conquer/simplebenders.jl [hunk 9]
```diff
@@ -1,15 +1,15 @@
-struct SimpleBenders <: AbstractConquerStrategy 
-    benderscutgen::BendersCutGeneration
-end
-SimpleBenders(;bendcutgen = BendersCutGeneration()) = SimpleBenders(bendcutgen)
+# struct SimpleBenders <: AbstractConquerStrategy
```
Comment (guimarqu) [inline src/Algorithm/strategies/conquer/simplebenders.jl]: rm file

File: src/Algorithm/reformulationsolver.jl [hunk 10]
```diff
@@ -1,221 +1,203 @@
...
-    pb = get_ip_primal_bound(node_incumbents)
-
-    print("Node bounds after treatment : ")
-    printbounds(db, pb)
-    println()
-
-    println("************************************************************")
-    return
-end
+
```
Comment (guimarqu) [inline src/Algorithm/reformulationsolver.jl]: rm file

File: src/Algorithm/masteripheur.jl [hunk 11]
```diff
@@ -1,39 +1,56 @@
...
-end
+# struct MasterIpHeuristicRecord <: AbstractAlgorithmResult
+#     incumbents::Incumbents
+# end
 
-function run!(algo::MasterIpHeuristic, form, node)
+# function prepare!(algo::MasterIpHeuristic, form, node)
+#     @logmsg LogLevel(-1) "Prepare MasterIpHeuristic."
+#     return
+# end
```
Comment (guimarqu) [inline src/Algorithm/masteripheur.jl]: rm

File: src/Algorithm/interface.jl [hunk 12]
```diff
@@ -1,72 +1,202 @@
...
+#         TO.@timeit Coluna._to "prepare" begin
+#             prepare!(algo, form, node)
+#         end
+#         TO.@timeit Coluna._to "run" begin
+#             record = run!(algo, form, node)
+#         end
+#     end
+#     set_algorithm_result!(node, algo, record)
+#     return record
+# end
```
Comment (guimarqu) [inline src/Algorithm/interface.jl]: rm comments

File: src/Algorithm/colgen.jl [hunk 13]
```diff
@@ -1,64 +1,87 @@
...
 
 # Overload of the algorithm's prepare function
-function prepare!(alg::ColumnGeneration, form, node)
-    @logmsg LogLevel(-1) "Prepare ColumnGeneration."
-    return
-end
+# function prepare!(alg::ColumnGeneration, form, node)
+#     @logmsg LogLevel(-1) "Prepare ColumnGeneration."
+#     return
+# end
```
Comment (guimarqu) [inline src/Algorithm/colgen.jl]: rm comments

File: src/Algorithm/Algorithm.jl [hunk 14]
```diff
@@ -21,38 +26,59 @@ const MOI = MathOptInterface
...
+# include("strategies/strategy.jl")
+
+# include("reformulationsolver.jl")
+
+# # Here include divide strategies
+# include("strategies/divide/simplebranching.jl") # to remove
 
 # Here include explore strategies
-include("strategies/explore/simplestrategies.jl")
+# include("strategies/explore/simplestrategies.jl")
```
Comment (guimarqu) [inline src/Algorithm/Algorithm.jl]: rm comments

File: src/Algorithm/benders.jl [hunk 15]
```diff
@@ -16,38 +16,56 @@ mutable struct BendersCutGenData
...
 
 # Overload of the solver interface
-function prepare!(algo::BendersCutGeneration, form, node)
-    @logmsg LogLevel(-1) "Prepare BendersCutGeneration."
-    return
-end
+# function prepare!(algo::BendersCutGeneration, form, node)
+#     @logmsg LogLevel(-1) "Prepare BendersCutGeneration."
+#     return
+# end
```
Comment (guimarqu) [inline src/Algorithm/benders.jl]: rm comments

File: test/unit/strategies/strategy.jl [hunk 16]
```diff
@@ -1,9 +1,9 @@
-function strategy_unit_tests()
-    strategy_fallbacks_tests()
-end
+# function strategy_unit_tests()
```
Comment (guimarqu) [inline test/unit/strategies/strategy.jl]: rm file

File: test/unit/unit_tests.jl [hunk 17]
```diff
@@ -57,9 +57,9 @@ function unit_tests()
     @testset "algorithm.jl" begin
         algorithm_unit_tests()
     end
-    @testset "strategy.jl" begin
-        strategy_unit_tests()
-    end
+    # @testset "strategy.jl" begin
+    #     strategy_unit_tests()
+    # end
```
Comment (guimarqu) [inline test/unit/unit_tests.jl]: rm comments

File: src/Algorithm/branching/branchingalgo.jl (hunk 1 above)
Comment (rrsadykov) [reply src/Algorithm/branching/branchingalgo.jl]: to remove line 92

**Pull request merged**